nemo/collections/asr/parts/context_biasing/boosting_graph_batched.py

```diff
@@ -0,0 +1,444 @@
+# Copyright (c) 2025, NVIDIA CORPORATION & AFFILIATES.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from collections import deque
+from dataclasses import InitVar, dataclass, field
+from typing import NamedTuple, Optional
+
+import numpy as np
+import torch
+from lightning.pytorch import Trainer
+from omegaconf import MISSING, DictConfig, OmegaConf
+from tqdm.auto import tqdm
+
+from nemo.collections.asr.parts.context_biasing.context_graph_universal import ContextGraph, ContextState
+from nemo.collections.asr.parts.submodules.ngram_lm import NGramGPULanguageModel
+from nemo.utils import logging
+
+
+class TBranch(NamedTuple):
+    """Structure (tuple) to represent a branch in the boosting tree"""
+
+    symbol: int  # token id
+    start_node: ContextState  # start node of the branch
+    next_node: ContextState  # next node of the branch
+
+
+@dataclass
+class BoostingTreeStorage:
+    """
+    NumPy-based storage for suffix tree (weighted acceptor) for phrase boosting
+    """
+
+    num_states_max: InitVar[int]
+    num_arcs_max: InitVar[int]
+
+    vocab_size: int
+    max_order: int
+
+    arcs: np.ndarray = field(init=False)
+    states: np.ndarray = field(init=False)
+
+    _node_cache: dict[int, int] = field(default_factory=dict)
+
+    unk_score: float = 0.0
+    final_eos_score: float = 0.0
+    num_states: int = 0
+    num_arcs: int = 0
+    start_state: int = 0
+    bos_state: int = 1
+
+    def __post_init__(self, num_states_max: int, num_arcs_max: int, separate_bos_state: bool = True):
```

Comment:
No need to use `separate_bos_state` here - it should be always False for boosting tree (True for n-gram LM)